File: Player.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ClassExtends = ____lualib.__TS__ClassExtends
local __TS__ArrayMap = ____lualib.__TS__ArrayMap
local __TS__ArraySplice = ____lualib.__TS__ArraySplice
local ____exports = {}
local ____Perk = require("Perk")
local Perk = ____Perk.default
local ____Card = require("Cards.Card")
local Card = ____Card.default
local ____Character = require("Character")
local Character = ____Character.default
local ____Dealer = require("Dealer")
local Dealer = ____Dealer.default
____exports.default = __TS__Class()
local Player = ____exports.default
Player.name = "Player"
__TS__ClassExtends(Player, Character)
function Player.prototype.____constructor(self, gameManager)
    Character.prototype.____constructor(self)
    self.gameManager = gameManager
    self.name = "Player"
    self.money = 0
    self.experience = 0
    self.level = 1
    self.discards = 3
    self.numberOfLootCards = 3
    self.perks = {}
end
function Player.prototype.load(self, data)
    self.name = data.name
    self.money = data.money
    self.experience = data.experience
    self.level = data.level
    self.discards = data.discards
    self.numberOfLootCards = data.numberOfLootCards
    self.hand = __TS__ArrayMap(
        data.hand,
        function(____, cardData) return Card:load(self.gameManager, cardData) end
    )
    self.deck = __TS__ArrayMap(
        data.deck,
        function(____, cardData) return Card:load(self.gameManager, cardData) end
    )
    self.discardPile = __TS__ArrayMap(
        data.discardPile,
        function(____, cardData) return Card:load(self.gameManager, cardData) end
    )
    self.perks = __TS__ArrayMap(
        data.perks,
        function(____, perkData) return Perk:load(self.gameManager, perkData) end
    )
end
function Player.prototype.save(self)
    return {
        name = self.name,
        money = self.money,
        experience = self.experience,
        level = self.level,
        discards = self.discards,
        numberOfLootCards = self.numberOfLootCards,
        hand = __TS__ArrayMap(
            self.hand,
            function(____, card) return card:save() end
        ),
        deck = __TS__ArrayMap(
            self.deck,
            function(____, card) return card:save() end
        ),
        discardPile = __TS__ArrayMap(
            self.discardPile,
            function(____, card) return card:save() end
        ),
        perks = __TS__ArrayMap(
            self.perks,
            function(____, perk) return perk:save() end
        )
    }
end
function Player.prototype.setup(self)
    if #self.deck == 0 then
        Dealer:initializePlayerDeck(self.gameManager)
    end
end
function Player.prototype.removeSelectedCardsFromHand(self)
    do
        local i = #self.hand - 1
        while i >= 0 do
            local card = self.hand[i + 1]
            if card.isSelected then
                card.isSelected = false
                self:addToDiscards(card)
                __TS__ArraySplice(self.hand, i, 1)
            end
            i = i - 1
        end
    end
end
function Player.prototype.discard(self)
    local newHand = {}
    for ____, card in ipairs(self.hand) do
        if card.isSelected then
            card.isSelected = false
            card:onUnselect()
            local ____self_discardPile_0 = self.discardPile
            ____self_discardPile_0[#____self_discardPile_0 + 1] = card
        else
            newHand[#newHand + 1] = card
        end
    end
    self.hand = newHand
end
function Player.prototype.anySelectedCards(self)
    for ____, card in ipairs(self.hand) do
        if card.isSelected then
            return true
        end
    end
    return false
end
function Player.prototype.cashout(self, points)
    if points < 0 then
        return
    end
    self.money = self.money + points
end
function Player.prototype.hasPerk(self, perkType)
    for ____, perk in ipairs(self.perks) do
        if perk.perkType == perkType then
            return true
        end
    end
    return false
end
function Player.prototype.addPerk(self, perk)
    local ____self_perks_1 = self.perks
    ____self_perks_1[#____self_perks_1 + 1] = perk
end
function Player.prototype.gatherExperience(self, exp)
    self.experience = self.experience + exp
    if self.experience >= self:getNextLevelExperience() then
        self:levelUp()
        return true
    end
    return false
end
function Player.prototype.getNextLevelExperience(self)
    repeat
        local ____switch38 = self.level
        local ____cond38 = ____switch38 == 1
        if ____cond38 then
            return 100
        end
        ____cond38 = ____cond38 or ____switch38 == 2
        if ____cond38 then
            return 150
        end
        ____cond38 = ____cond38 or ____switch38 == 3
        if ____cond38 then
            return 250
        end
        ____cond38 = ____cond38 or ____switch38 == 4
        if ____cond38 then
            return 500
        end
        do
            return 1000
        end
    until true
end
function Player.prototype.levelUp(self)
    self.experience = 0
    self.level = self.level + 1
end
function Player.prototype.unselectCards(self)
    for ____, card in ipairs(self.hand) do
        if card.isSelected then
            card.isSelected = false
            card:onUnselect()
        end
    end
end
return ____exports

File: Player.ts
/** @noSelfInFile */

import Card, { CardData } from "Cards/Card";
import Dealer from "Dealer";
import { CharacterTypes, Perks, TextIds } from "Enums";
import GameManager from "GameManager";
import Perk, { PerkData } from "Perk";
import Character from "./Character";

interface PlayerData {
  name: string;
  money: number;
  experience: number;
  level: number;
  discards: number;
  numberOfLootCards: number;
  hand: CardData[];
  deck: CardData[];
  discardPile: CardData[];
  perks: PerkData[];
}

export default class Player extends Character {
  name = "Player";
  money = 0;
  experience = 0;
  level = 1;
  discards = 3;
  numberOfLootCards = 3;
  perks: Perk[] = [];

  constructor(gameManager: GameManager) {
    super(gameManager, CharacterTypes.PLAYER);
  }

  load(data: PlayerData): void {
    this.name = data.name;
    this.money = data.money;
    this.experience = data.experience;
    this.level = data.level;
    this.discards = data.discards;
    this.numberOfLootCards = data.numberOfLootCards;
    this.hand = data.hand.map((cardData) =>
      Card.load(this.gameManager, cardData)
    );
    this.deck = data.deck.map((cardData) =>
      Card.load(this.gameManager, cardData)
    );
    this.discardPile = data.discardPile.map((cardData) =>
      Card.load(this.gameManager, cardData)
    );
    this.perks = data.perks.map((perkData) =>
      Perk.load(this.gameManager, perkData)
    );
  }

  save(): PlayerData {
    return {
      name: this.name,
      money: this.money,
      experience: this.experience,
      level: this.level,
      discards: this.discards,
      numberOfLootCards: this.numberOfLootCards,
      hand: this.hand.map((card) => card.save()),
      deck: this.deck.map((card) => card.save()),
      discardPile: this.discardPile.map((card) => card.save()),
      perks: this.perks.map((perk) => perk.save()),
    };
  }

  setup(): void {
    if (this.deck.length === 0) {
      Dealer.initializePlayerDeck(this.gameManager);
    }
  }

  anySelectedCards(): boolean {
    for (const card of this.hand) {
      if (card.isSelected) {
        return true;
      }
    }
    return false;
  }

  getSelectedCards(): Card[] {
    const selectedCards: Card[] = [];
    for (const card of this.hand) {
      if (card.isSelected) {
        selectedCards.push(card);
      }
    }
    return selectedCards;
  }

  cashout(points: number): void {
    if (points < 0) return;
    this.addMoney(points);
  }

  addMoney(amount: number): void {
    this.money += amount;
    this.gameManager.assetManager.textManager.updateText(
      TextIds.PLAYER_PORTRAIT_MONEY,
      `${this.money}`
    );
  }

  hasPerk(perkType: Perks): boolean {
    for (const perk of this.perks) {
      if (perk.perkType === perkType) {
        return true;
      }
    }
    return false;
  }

  addPerk(perk: Perk): void {
    this.perks.push(perk);
  }

  gatherExperience(exp: number): boolean {
    this.addExperience(exp);

    if (this.experience >= this.getNextLevelExperience()) {
      this.levelUp();
      return true;
    }
    return false;
  }

  getNextLevelExperience(): number {
    switch (this.level) {
      case 1:
        return 100;
      case 2:
        return 150;
      case 3:
        return 250;
      case 4:
        return 500;
      default:
        return 1000;
    }
  }

  levelUp(): void {
    this.experience = 0;
  }

  addExperience(exp: number): void {
    this.experience += exp;
    this.gameManager.assetManager.textManager.updateText(
      TextIds.PLAYER_PORTRAIT_EXPERIENCE,
      `${this.experience} XP`
    );
  }

  unselectCards(): void {
    for (const card of this.hand) {
      if (card.isSelected) {
        card.onUnselect();
      }
    }
  }

  discard(): number[] {
    const newHand: Card[] = [];
    const removedIndices: number[] = [];
    for (let index = 0; index < this.hand.length; index++) {
      const card = this.hand[index];
      if (card.isSelected) {
        card.onUnselect();
        this.addToDiscards(card);
        removedIndices.push(index);
      } else {
        newHand.push(card);
      }
    }
    this.hand = newHand;
    return removedIndices;
  }
}
